base/bitset.jl

```diff
@@ -1,112 +1,237 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-struct BitSet <: AbstractSet{Int}
-    bits::BitVector
-    BitSet() = new(sizehint!(falses(0), 256))
+const Bits = Vector{UInt64}
+const Chk0 = zero(UInt64)
+const NoOffset = -one(Int) << 60
+# + NoOffset must be big enough to stay < 0 when add with any offset.
+#   An offset is in the range -2^57:2^57
+# + when the offset is NoOffset, the bits field *must* be empty
+# + NoOffset could be made to be > 0, but a negative one allows
+#   a small optimization in the in(x, ::BitSet)
+
+mutable struct BitSet <: AbstractSet{Int}
+    bits::Vector{UInt64}
+    # 1st stored Int equals 64*offset
+    offset::Int
+
+    BitSet() = new(sizehint!(zeros(UInt64, 0), 4), NoOffset)
 end
 
 """
     BitSet([itr])
 
-Construct a sorted set of positive `Int`s generated by the given iterable object, or an
-empty set. Implemented as a bit string, and therefore designed for dense integer sets. Only
-`Int`s greater than 0 can be stored. If the set will be sparse (for example holding a few
+Construct a sorted set of `Int`s generated by the given iterable object, or an
+empty set. Implemented as a bit string, and therefore designed for dense integer sets.
+If the set will be sparse (for example holding a few
 very large integers), use [`Set`](@ref) instead.
 """
 BitSet(itr) = union!(BitSet(), itr)
 
+@inline intoffset(s::BitSet) = s.offset << 6
+
 eltype(::Type{BitSet}) = Int
 similar(s::BitSet) = BitSet()
 copy(s1::BitSet) = copy!(BitSet(), s1)
 function copy!(dest::BitSet, src::BitSet)
     resize!(dest.bits, length(src.bits))
     copy!(dest.bits, src.bits)
+    dest.offset = src.offset
     dest
 end
+
 eltype(s::BitSet) = Int
-sizehint!(s::BitSet, n::Integer) = (n > length(s.bits) && _resize0!(s.bits, n); s)
 
-# An internal function for setting the inclusion bit for a given integer n >= 0
-@inline function _setint!(s::BitSet, idx::Integer, b::Bool)
-    if idx > length(s.bits)
+sizehint!(s::BitSet, n::Integer) = sizehint!(s.bits, n)
+
+# given an integer i, return the chunk which stores it
+chk_indice(i::Int) = i >> 6
```

Comment:
This and the next are the same as `_div64` and `_mod64` from "bitarray.jl". I know those names are not great, but it seems slightly silly to have two sets of nearly identical definitions.

Reply:
Ah yeah right! They are like that because they evolved: first they did what `get_chunks_id` do, then they were updated to work for all `Int` range by promoting to `Int128`, and ended up this way; but `chk_indice` must use the signed shift operator, so can `_div64` can be changed to use `>>` ?

Reply:
Yes, no part of the code which uses `_div64` is ever supposed to get a negative integer, and it would give meaningless results anyway if it did, so  `>>` should be fine.

Reply:
Ok thanks, done